### PR TITLE
[WIP] gru.py and related scripts

### PR DESCRIPTION
**Pull request by GaofengCheng**
@danpovey 
Hi Dan, I have updated the gru.py, and run locally. Compared with my previously generated configs, this python script can generate the same configs as those in my previous experiments.

I'm preparing the bash scripts under SWBD and AMI, and will update them under this PR.

### CODE REVIEW

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+import sys
+from libs.nnet3.xconfig.basic_layers import XconfigLayerBase
+
+# This class is for lines like
+#   'gru-layer name=gru1 input=[-1] delay=-3'
+# It generates an GRU sub-graph without output projections.
+# The output dimension of the layer may be specified via 'cell-dim=xxx', but if not specified,
+# the dimension defaults to the same as the input.
+# See other configuration values below.
+# decay-time is deprecated under GRU or PGRU, as I found the PGRUs do not need the decay-time option to get generalized to unseen sequence length
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: please remove decay-time before we check this in.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+        configs.append("component name={0}.W_h.UW type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + cell_dim, cell_dim , affine_str))
+        
+        configs.append("# Defining the non-linearities")
+        configs.append("component name={0}.z type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.r type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.h type=TanhComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+
+        configs.append("# Defining the components for other cell computations")
+        configs.append("component name={0}.h1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: When do you plan to add the 'fast' versions of these layers, that use the compoennts I added?

Comment (GaofengCheng) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: @danpovey right now

**Comment by francisr** [comment]
What's the difference between gru and fast-gru?

**Comment by GaofengCheng** [comment]
@francisr same thing, but fast version has an accelerated kernel. search "GruNonlinearityComponent" for the detail of fast-gru

**Comment by francisr** [comment]
So why does it consistently have worse performances?

**Comment by GaofengCheng** [comment]
@francisr fast experiments were tuned by @hlthu , I'm checking his fast version results

**Comment by GaofengCheng** [comment]
@francisr the tdnn-pgru on my env is OK, I updated the new scripts. Anyway, tdnn-opgru is recommended.

**Comment by danpovey** [comment]
If you are now recommending the OPGRU always in preference to PGRU, and that is not what is implemented in the GruNonlinearityComponent, I see no reason to keep the GruNonlinearityComponent code.
Of course if you could implement a version that would be helpful for OPGRU that would be nice.  But I'd be OK to merge a version without the special component, with just the scripts.
What do you think?

**Comment by GaofengCheng** [comment]
@danpovey I agree with you, deleting GruNonlinearityComponent now, and I will later make a PR about the fast-version of OPGRU.

**Comment by GaofengCheng** [comment]
This will soon replace by a new PR. fast version and the normal version component are separated.

**Comment by danpovey** [comment]
@vimalmanohar, do you have time to check through this for any minor problems?

**Comment by GaofengCheng** [comment]
@vimalmanohar PR #1950 is the final PR for merging

**Comment by vimalmanohar** [comment]
Ok, I'll have a look at it.

**Pull request closed without merging**